### PR TITLE
Split the inheirtance on the RequestConditions

### PR DESCRIPTION
**Pull request by RickWinter**
Rename RequestConditions to ModifiedConditions

Change splits apart the RequestConditions to conditions based on ETags `MatchConditions` and conditions based on time `ModifiedConditions`.

# Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/master/CONTRIBUTING.md#pull-requests).

- [x] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [x] Doxygen docs
- [x] Unit tests
- [x] No unwanted commits/changes
- [x] Descriptive title/description
  - [x] PR is single purpose
  - [x] Related issue listed
- [x] Comments in source
- [x] No typos
- [ ] Update changelog
- [x] Not work-in-progress
- [x] External references or docs updated
- [x] Self review of PR done
- [x] Any breaking changes?

### CODE REVIEW

**Review by JeffreyRichter — APPROVED** [review]

**Comment by ahsonkhan** [comment]
@RickWinter - is this for Feb?

File: sdk/core/azure-core/inc/azure/core/modified_conditions.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/modified_conditions.hpp:22]: Needs changelog update.

File: sdk/core/azure-core/inc/azure/core/modified_conditions.hpp
Comment (ahsonkhan) [inline sdk/core/azure-core/inc/azure/core/modified_conditions.hpp:20]: What motivated this change, given some of the other languages like .NET do it with the inheritance and call it `RequestConditions`?

https://github.com/Azure/azure-sdk-for-net/blob/b493f7ad5487fd0d01f004e9f7ebdf50bfec0b96/sdk/core/Azure.Core/src/RequestConditions.cs#L11

**Review by ahsonkhan — CHANGES_REQUESTED** [review]
Need CL update, and blocking if it isn't meant for Feb (given storage likely would need to react to it).

**Comment by Jinming-Hu** [comment]
`IfModifiedSince` and `IfUnmodifiedSince` need to be nullable.

File: sdk/core/azure-core/inc/azure/core/modified_conditions.hpp
Comment (Jinming-Hu) [reply sdk/core/azure-core/inc/azure/core/modified_conditions.hpp:20]: Jeffrey said yes. https://github.com/Azure/azure-sdk-for-cpp/issues/1608

Comment (ahsonkhan) [reply sdk/core/azure-core/inc/azure/core/modified_conditions.hpp:20]: Thanks for linking to the issue.

**Review by vhvb1989 — APPROVED** [review]
Update Changelog before merging

**Comment by RickWinter** [comment]
> Update Changelog before merging

Yep, was waiting for all the release before adding that.  Didn't want the merge conflict :)

**Review by ahsonkhan — APPROVED** [review]

**Comment by ahsonkhan** [comment]
Fixes https://github.com/Azure/azure-sdk-for-cpp/issues/1608

**Pull request merged**
